prov/util/src/util_uffd.c

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright (c) 2019 Amazon.com, Inc. or its affiliates. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <ofi_util.h>
+
+#ifdef HAVE_LINUX_USERFAULTFD_H
+#include <config.h>
+#include <ofi_uffd.h>
+#include <errno.h>
+#include <sys/syscall.h>
+#include <sys/ioctl.h>
+#include <linux/userfaultfd.h>
+
+#define MB(x)((size_t)(x) << 20)
+#define GB(x)((size_t)(x) << 30)
+#define ARRAY_SIZE(A) (sizeof(A)/sizeof(*A))
+
+/* See TODO in function ofi_uffd_register */
+static long page_sizes[] = {4096, MB(2), GB(1)};
+
+static inline void set_uffdio_range(struct uffdio_range *range,
+				    void *start, size_t len,
+				    long page_size)
+{
+	range->start = ((uint64_t)start / page_size) * page_size;
+	range->len = ofi_div_ceil(len, page_size) * page_size;
+}
```

Comment:
I believe the above is incorrect.  If start is not page aligned, an extra page may be needed.  The referenced patch contains helper functions that I plan on using for this:

https://github.com/ofiwg/libfabric/pull/4956/commits/b15c579e1cda4ebbcec8ce6eb8734aa50816a0e2

It would be good to get a review on those changes, as they are critical to get the correct ranges.

Reply:
If start is not page aligned, it will round down to the nearest page. I will review your patch

Reply:
My concern is that len is calculated incorrectly.